qui/tray/devices.py

```diff
@@ -132,321 +66,267 @@ def attach_item(self, menu_item):
+    """ MenuItem showing the device data and a :class:`DomainMenu`. """
 
-    def __init__(self, device, frontend_domains, qapp,
-                 dispatcher, gtk_app, *args, **kwargs):
-        "docstring"
+    def __init__(self, device, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        self.gtk_app = gtk_app
-        self.qapp = qapp
         self.device = device
-        self.devclass = self.device.devclass
-        self.frontend_domains = frontend_domains
-        self.dispatcher = dispatcher
-        vm_icon = self.device.backend_domain.label.icon
-        self.hbox = qui.decorators.device_hbox(
-            self.device,
-            frontend_domains=self.frontend_domains)  # type: Gtk.Box
 
-        self.set_image(qui.decorators.create_icon(vm_icon))
+        self.hbox = qui.decorators.device_hbox(self.device)  # type: Gtk.Box
+
+        self.set_image(qui.decorators.create_icon(self.device.vm_icon))
+
         self.add(self.hbox)
-        submenu = DomainMenu(
-            self.device, self.frontend_domains, qapp, self.dispatcher,
-            self.gtk_app)
-        self.set_submenu(submenu)
 
-        self.dispatcher.add_handler('domain-shutdown',
-                                    self.vm_shutdown)
-        self.dispatcher.add_handler('domain-start-failed',
-                                    self.vm_shutdown)
-        self.dispatcher.add_handler('domain-start', self.vm_start)
 
-    def vm_start(self, vm, _event, **_kwargs):
-        if self.device in vm.devices[self.devclass].attached():
-            self.device_attached(vm)
+class Device:
+    def __init__(self, dev):
+        self.dev_name = str(dev)
+        self.ident = dev.ident
+        self.description = dev.description
+        self.devclass = dev.devclass
+        self.attachments = set()
+        self.backend_domain = dev.backend_domain.name
+        self.vm_icon = dev.backend_domain.label.icon
 
-    def vm_shutdown(self, vm, _event, **_kwargs):
-        if vm in self.frontend_domains:
-            self.device_detached(vm)
-
-    def device_attached(self, vm):
-        self.frontend_domains.append(vm)
-        self.remove(self.hbox)
-        self.hbox = qui.decorators.device_hbox(
-            self.device, frontend_domains=self.frontend_domains)
-        self.add(self.hbox)
-        self.get_submenu().dev_attached(vm)
-        self.show_all()
-
-    def device_detached(self, vm):
-        if vm:
-            self.frontend_domains.remove(vm)
-        self.remove(self.hbox)
-        self.hbox = qui.decorators.device_hbox(
-            self.device, frontend_domains=self.frontend_domains)
-        self.add(self.hbox)
-        self.get_submenu().dev_detached(vm)
-        self.show_all()
+    def __str__(self):
+        return self.dev_name
 
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+
+class VM:
+    def __init__(self, vm):
+        self.hash = hash(vm)
```

Comment:
This internal attribute, must not be modified from outside and preferably also accessed. Please rename to `__hash`.